### PR TITLE
[xpu][feature] Enable triton online softmax kernels on XPU.

### PR DESCRIPTION
**Pull request by weishi-deng**
This pr is to enable triton online softmax kernels for xpu devices, so we add a device check in prepare_softmax_extra_check.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

### CODE REVIEW

File: torch/_inductor/fx_passes/post_grad.py [hunk 1]
```diff
@@ -322,8 +322,10 @@ def prepare_softmax_extra_check(match):
     """
     return (
         config.online_softmax
-        and match.kwargs["x"].meta["val"].device.type == "cuda"
-        and config.cuda_backend == "triton"
+        and (match.kwargs["x"].meta["val"].device.type == "cuda"
```
Comment (etaf) [inline torch/_inductor/fx_passes/post_grad.py]: Suggest to make the logic more clear:
```        
and (
   (match.kwargs["x"].meta["val"].device.type == "cuda" and config.cuda_backend == "triton")         
or (match.kwargs["x"].meta["val"].device.type == "xpu" and config.xpu_backend == "triton")
)
```

**Comment by etaf** [comment]
Is there any cuda test case we can reuse for the online-softmax?  Please enable the corresponding UT.

File: torch/_inductor/fx_passes/post_grad.py (hunk 1 above)
Comment (weishi-deng) [reply torch/_inductor/fx_passes/post_grad.py]: Fixed

**Comment by etaf** [comment]
> > Is there any cuda test case we can reuse for the online-softmax? Please enable the corresponding UT.
> 
> The ut for online softmax on xpu is enabled in test_online_softmax.py. Note that the split reduction heuristic is relative to the device hardware info, so the output code with a shape of 2048 on xpu devices is different from that of CUDA. BMG has num_sm of 20, so the check in https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L1362 (as 2048 > 20 * 2 *32) will make num_split 1.

Great! could you add these explanations as comments in the test case so that others are aware of the differences from CUDA?

**Comment by weishi-deng** [comment]
> > > Is there any cuda test case we can reuse for the online-softmax? Please enable the corresponding UT.
> > 
> > 
> > The ut for online softmax on xpu is enabled in test_online_softmax.py. Note that the split reduction heuristic is relative to the device hardware info, so the output code with a shape of 2048 on xpu devices is different from that of CUDA. BMG has num_sm of 20, so the check in https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L1362 (as 2048 > 20 * 2 *32) will make num_split 1.
> 
> Great! could you add these explanations as comments in the test case so that others are aware of the differences from CUDA?

Done

**Review by etaf — APPROVED** [review]

File: torch/_inductor/fx_passes/post_grad.py [hunk 2]
```diff
@@ -322,8 +322,10 @@ def prepare_softmax_extra_check(match):
     """
     return (
         config.online_softmax
-        and match.kwargs["x"].meta["val"].device.type == "cuda"
-        and config.cuda_backend == "triton"
+        and (
+            (match.kwargs["x"].meta["val"].device.type == "cuda" and config.cuda_backend == "triton")
+            or (match.kwargs["x"].meta["val"].device.type == "xpu" and config.xpu_backend == "triton")
```
Comment (EikanWang) [inline torch/_inductor/fx_passes/post_grad.py]: Could it be something like this?
```python
match.kwargs["x"].meta["val"].device.type in ["cuda", "xpu"] and "triton" in [config.cuda_backend, config.xpu_backend]
```

File: test/inductor/test_online_softmax.py [hunk 3]
```diff
@@ -310,5 +312,5 @@ def f(x, y):
 instantiate_parametrized_tests(TestOnlineSoftmax)
 
 if __name__ == "__main__":
-    if IS_LINUX and HAS_CUDA_AND_TRITON:
+    if IS_LINUX and (HAS_CUDA_AND_TRITON or HAS_XPU_AND_TRITON):
```
Comment (EikanWang) [inline test/inductor/test_online_softmax.py]: Could it be something like `HAS_TRITON` and `HAS_GPU`?

**Review by EikanWang — APPROVED** [review]
Pls. rebase this PR.

**Review by mlazos — APPROVED** [review]

**Comment by weishi-deng** [comment]
@EikanWang @etaf I updated the ut to add the check with device info as the split reduction is enabled on XPU PVC but disabled on XPU BMG devices. Therefore, we should add a condition check to determine when to use split reduction. Pls help review again.

File: torch/_inductor/fx_passes/post_grad.py [hunk 4]
```diff
@@ -351,10 +351,9 @@ def prepare_softmax_extra_check(match):
     """
     We only have triton online softmax kernels currently.
     """
-    return (
-        config.online_softmax
-        and match.kwargs["x"].meta["val"].device.type == "cuda"
-        and config.cuda_backend == "triton"
+    return config.online_softmax and (
```
Comment (etaf) [inline torch/_inductor/fx_passes/post_grad.py]: from torch._inductor.utils import get_current_backend
device_type = match.kwargs["x"].meta["val"].device.type
return config.online_softmax and device_type in ("cuda", "xpu") and "triton" == get_current_backend(device_type)

**Review by etaf — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/post_grad.py (hunk 4 above)
Comment (weishi-deng) [reply torch/_inductor/fx_passes/post_grad.py]: Updated.

File: torch/_inductor/fx_passes/post_grad.py [hunk 5]
```diff
@@ -351,11 +351,9 @@ def prepare_softmax_extra_check(match):
     """
     We only have triton online softmax kernels currently.
     """
-    return (
-        config.online_softmax
-        and match.kwargs["x"].meta["val"].device.type == "cuda"
-        and config.cuda_backend == "triton"
-    )
+    from torch._inductor.utils import get_current_backend
```
Comment (etaf) [inline torch/_inductor/fx_passes/post_grad.py]: Oh, I'm not mean to put the import line here, I just try to tell where you can get the `get_current_backend`. it'll be better to move the import line to the file header.

**Review by etaf — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/post_grad.py (hunk 5 above)
Comment (weishi-deng) [reply torch/_inductor/fx_passes/post_grad.py]: updated.

**Review by etaf — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Comment by weishi-deng** [comment]
> @weishi-deng would you mind taking a look at #172048 I would appreciate your input on the changes I made. Should I have kept the fallback to num_kernels = 1 for XPUs? Thanks

I made a test for #172048, and it also works for XPU devices with/without num_sm > 20. I think this looks good to me.
